### ISSUE TITLE
OTLP exemplar support

### ISSUE THREAD

**Issue by shakuzen** [micrometer-metrics/micrometer#3795]
Similar to the exemplar support we added for the Prometheus registry, we can add exemplar support for the OTLP registry. See https://opentelemetry.io/docs/reference/specification/metrics/data-model/#exemplars

**Comment by 0x006EA1E5** [comment]
Hi,

Is there are movement on this?

Thanks

**Comment by sonallux** [comment]
Hi @shakuzen and @jonatan-ivanov, is someone already working on this? We stumpled about this missing feature of the OTLP micrometer registry and I would like to work on this.

**Comment by shakuzen** [comment]
I believe @jonatan-ivanov has started work on this, but I'll let him give any update or say if he would like help with it.

**Comment by jonatan-ivanov** [comment]
@sonallux Thank you for the offer! If you want to work on it, please feel free to do so but there is a catch. :) There are multiple different ways to implement this and I'm right now still exploring them. If you have thoughts on how you would implement this or want to create a draft/prototype, please don't hesitate to share, we are appreciating inputs on this.
It is not guaranteed that we will chose the direction you did but feedback is welcomed. I'm only saying this because I don't want you to be disappointed if you work a lot on a PR and we go to a different direction but it can also happen that we will apply the code you come up with or do something influenced by you. Does this sound fair?
I will be in and out in the next two weeks due to conferences so might not be very responsive in the near future but please feel free to share what you have. Other than this issue, you can also create a room with Tommy (`@Tommy Ludwig`) and me (`@Jonatan Ivanov`) on https://slack.micrometer.io if you want to discuss/share anything.

**Comment by sonallux** [comment]
> It is not guaranteed that we will chose the direction you did but feedback is welcomed. I'm only saying this because I don't want you to be disappointed if you work a lot on a PR and we go to a different direction but it can also happen that we will apply the code you come up with or do something influenced by you. Does this sound fair?

@jonatan-ivanov Yes, that is okay for me. I have already created a working prototype (see PR #5486). Please let me know, what you think about this draft.

**Comment by Nikzz3** [comment]
I would also love to have this feature.

Is there already an estimated time for release?

Thanks.

**Comment by ryleighmikami** [comment]
Hello team, upvoting this

**Comment by illenko** [comment]
Waiting for this feature.

Thanks!

**Comment by ThomasVitale** [comment]
Is there anything I can do to help with this?

**Comment by vmeunier** [comment]
Hello
It's almost 3 years old now, every framework supports exemplars but micrometer... It's clearly a huge piece of observability missing.

**Comment by jonatan-ivanov** [comment]
We have a PR that should close this: https://github.com/micrometer-metrics/micrometer/pull/7199
Please feel free to take a look, try out, and give feedback.

**Comment by shakuzen** [comment]
This is available in snapshots now. Please try it out and give feedback. It will also be in the 1.17.0-M3 milestone release next week. We'd really appreciate feedback on it before the RC1 release next month. See #7199 or the [new docs](https://docs.micrometer.io/micrometer/reference/1.17-SNAPSHOT/implementations/otlp.html#_exemplars) for more information on how to configure it.

**Comment by vmeunier** [comment]
It's working fine with 1.17.0-SNAPSHOT with the configuration provided for a SpringBoot app.

I have a question though : I thought examplars would have traces that are close to the value of the percentile they are bound to. From what I'm testing it's not working like that and seems more like random traces are being picked up. Am I missing something or is it intented to be that way ?

**Comment by jonatan-ivanov** [comment]
Are you using client-side percentiles (`.publishPercentiles(...)`), or histograms (`.publishPercentileHistogram()`) and calculate percentiles on the backend?

For the former, we don't provide exemplars, for the latter, we have exemplars per bucket so (depending on the certain case) Micrometer should publish exemplars from that neighborhood.
There are a few caveats though:
- We publish one exemplar per bucket per minute
- When I tried things with the OTel collector and Prometheus (and visualized things on a Grafana heat-map, I saw sparse exemplars so something is dropping them)

**Comment by vmeunier** [comment]
I'm using SpringBoot and tell it to publish percentile histograms.

Since i'm using [grafana/otel-lgtm](https://github.com/grafana/docker-otel-lgtm) docker container maybe I'm facing the same issue you're mentioning `When I tried things with the OTel collector and Prometheus (and visualized things on a Grafana heat-map, I saw sparse exemplars so something is dropping them)`. I added logs to the otlp exporter and I can see all the exemplars being pushed every minute... Yet I still don't see each of them on Grafana.

**Comment by jonatan-ivanov** [comment]
I think it might worth opening an issue to grafana with a question, please let us know if you do so.
